File: models/Action.ts
import ActionParams from '../interfaces/ActionParams';

class Action {
  private type: string;
  private params: ActionParams = {};

  constructor (type: string, params: ActionParams = {}) {
    this.type = type;
    this.params = params;
  }

  getType(): string { return this.type }

  getParams(): ActionParams { return this.params }

  getParam(param_name: string, default_value: any = undefined): any {
    return this.params[param_name] ?? default_value
  }
}

export default Action;

File: rust/models/Action.rs
use crate::interfaces::ActionParams;

pub struct Action {
    action_type: String,
    params: ActionParams,
}

impl Action {
    pub fn new(action_type: String, params: ActionParams) -> Self {
        Action { action_type, params }
    }

    pub fn get_type(&self) -> &String {
        &self.action_type
    }

    pub fn get_params(&self) -> &ActionParams {
        &self.params
    }

    pub fn get_param(&self, param_name: &str, default_value: Option<&str>) -> Option<&str> {
        self.params.get(param_name).or(default_value)
    }
}
